overlord/ifacestate/apparmorprompting/prompting.go

```diff
@@ -0,0 +1,551 @@
+	}
+	defer func() {
+		if retErr != nil {
+			promptsBackend.Close()
+		}
+	}()
+
+	rulesBackend, err := requestrules.New(notifyRule)
+	if err != nil {
+		return nil, fmt.Errorf("cannot open request rules backend: %w", err)
+	}
+	defer func() {
+		if retErr != nil {
+			rulesBackend.Close()
+		}
+	}()
+
+	m = &InterfacesRequestsManager{
+		listener:     listenerBackend,
+		prompts:      promptsBackend,
+		rules:        rulesBackend,
+		notifyPrompt: notifyPrompt,
+		notifyRule:   notifyRule,
+	}
+
+	m.tomb.Go(m.run)
+
+	return m, nil
+}
+
+// Run is the main run loop for the manager, and must be called using tomb.Go.
+func (m *InterfacesRequestsManager) run() error {
+	m.tomb.Go(func() error {
+		logger.Noticef("starting listener")
+		if err := listenerRun(m.listener); err != listener.ErrClosed {
+			return err
+		}
+		return nil
+	})
+
+	logger.Noticef("ready for prompts")
+run_loop:
+	for {
+		logger.Debugf("waiting prompt loop")
+		select {
+		case req, ok := <-listenerReqs(m.listener):
+			if !ok {
+				// Reqs() closed, so either errored or Stop() was called.
+				// In either case, the listener Close() method has already
+				// been called, and the tomb error will be set to the return
+				// value of the Run() call from the previous tracked goroutine.
+				logger.Noticef("listener closed requests channel")
+				return m.disconnect()
+			}
+			// XXX: this debug log leaks information about internal activity
+			logger.Debugf("Got from kernel req chan: %v", req)
+			if err := m.handleListenerReq(req); err != nil { // no use multithreading, since IsPathAllowed locks
+				// XXX: this log leaks information about internal activity
+				logger.Noticef("Error while handling request: %+v", err)
+			}
+		case <-m.tomb.Dying():
+			logger.Noticef("InterfacesRequestsManager tomb is dying, disconnecting")
+			break run_loop
+		}
+	}
+	return m.disconnect()
+}
+
+func (m *InterfacesRequestsManager) handleListenerReq(req *listener.Request) error {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	userID := uint32(req.SubjectUID)
+	// TODO: immediately deny if req.SubjectUID() == 0 (root)
+	snap := req.Label // Default to apparmor label, in case process is not a snap
+	tag, err := naming.ParseSecurityTag(req.Label)
+	if err == nil {
+		// the triggering process is not a snap, so treat apparmor label as snap field
+		snap = tag.InstanceName()
+	}
+
+	// TODO: when we support interfaces beyond "home", do a proper selection here
+	iface := "home"
+
+	path := req.Path
+
+	permissions, err := prompting.AbstractPermissionsFromAppArmorPermissions(iface, req.Permission)
+	if err != nil {
+		// XXX: this log leaks information about internal activity
+		logger.Noticef("error while parsing AppArmor permissions: %v", err)
+		response := listener.Response{
+			Allow:      false,
+			Permission: req.Permission,
+		}
+		return requestReply(req, &response)
+	}
+
+	remainingPerms := make([]string, 0, len(permissions))
```

Comment:
I think you only need to take the lock around here or do replies need to be ordered as well?